packages/signers/ledger-signer/src/ledger-signer.ts

```diff
@@ -0,0 +1,334 @@
+import type Transport from "@ledgerhq/hw-transport"
+import { merkleizeMetadata } from "@polkadot-api/merkleize-metadata"
+import type { PolkadotSigner } from "@polkadot-api/polkadot-signer"
+import {
+  Binary,
+  compact,
+  enhanceEncoder,
+  u16,
+  u32,
+  u8,
+} from "@polkadot-api/substrate-bindings"
+import { mergeUint8 } from "@polkadot-api/utils"
+import { getMetadata } from "./get-metadata"
+import { CLA, DEFAULT_SS58, INS, P1, P2 } from "./consts"
+
+const METADATA_IDENTIFIER = "CheckMetadataHash"
+const versionCodec = enhanceEncoder(
+  u8.enc,
+  (value: { signed: boolean; version: number }) =>
+    (+!!value.signed << 7) | value.version,
+)
+
+// 44'/354'
+const START_PATH = Uint8Array.from([44, 0, 0, 128, 98, 1, 0, 128])
+// 0'
+const MID_PATH = Uint8Array.from([0, 0, 0, 128])
+const HARDENED = 0x80000000 // 1 << 31
+const encodePath = (path1: number, path2: number) => {
+  // ensure numbers are positive integers and fit in 31 bits
+  if (
+    !Number.isInteger(path1) ||
+    path1 < 0 ||
+    path1 >= HARDENED ||
+    !Number.isInteger(path2) ||
+    path2 < 0 ||
+    path2 >= HARDENED
+  )
+    throw new Error(`Invalid path segments ${path1}, ${path2}`)
+
+  return mergeUint8(
+    START_PATH,
+    u32.enc(HARDENED + path1),
+    MID_PATH,
+    u32.enc(HARDENED + path2),
+  )
+}
+
+const getPubkeyMapKey = (path1: number, path2: number): string => {
+  return `${path1}:${path2}`
+}
+
+const [preBytes, postBytes] = ["<Bytes>", "</Bytes>"].map((str) =>
+  Binary.fromText(str).asBytes(),
+)
+
+/**
+ * ATTENTION: This class requires `Buffer` to be available. This is an Ledger
+ * requirement that we need to fulfill. If you are on a browser-based
+ * environment, make sure you polyfill it.
+ */
+export class LedgerSigner {
+  #transport: Transport
+  #pubkeys: Map<string, Uint8Array> // `${path1}:${path2}`
+  #deviceId: number | undefined
```

Comment:
These are essentially caches, right? Can't this be problematic if the user swaps a device?

Maybe an alternative API that would make this clearer is if instead of having a "LedgerSigner", you have something that gives you a "LedgerDevice" (or something like that) which already has the keys prepopulated, and if the user were to remove that device then that becomes invalid.

I'm missing context about how ledger works so maybe what I'm suggesting doesn't make sense.

Reply:
These are caches, right. Your concern makes sense, but the constructor of the `LedgerSigner` class takes a transport that the consumer created in advance. The transport uniquely identifies one device (usually a HID address).